independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java

```diff
@@ -195,24 +191,28 @@ public void destroy(ContextState state) {
         }
         if (state instanceof RequestContextState) {
             RequestContextState reqState = ((RequestContextState) state);
-            reqState.isValid = false;
-            synchronized (state) {
-                Map<Contextual<?>, ContextInstanceHandle<?>> map = ((RequestContextState) state).map;
+            if (reqState.isValid) {
+                reqState.isValid = false;
                 // Fire an event with qualifier @BeforeDestroyed(RequestScoped.class) if there are any observers for it
                 try {
                     fireIfNotEmpty(beforeDestroyedNotifier);
                 } catch (Exception e) {
                     LOG.warn("An error occurred during delivery of the @BeforeDestroyed(RequestScoped.class) event", e);
                 }
-                //Performance: avoid an iterator on the map elements
-                map.forEach(this::destroyContextElement);
-                // Fire an event with qualifier @Destroyed(RequestScoped.class) if there are any observers for it
+                synchronized (state) {
```

Comment:
I wonder, why we synchronize over `state` here? If there's possibly multiple threads calling `destroy(ContextState)`, then the `synchronized` block should be as before, otherwise we risk firing the `@BeforeDestroyed` / `@Destroyed` events multiple times.

Reply:
> otherwise we risk firing the @BeforeDestroyed / @Destroyed events multiple times.

Unfortunately, that was the case before this change because we did not check the `RequestContextState#isValid` flag. So even if we synchronized over `state` the events would be fired anyway. The truth is that the current code is not 100% correct either because `RequestContextState#isValid` is `volatile` and so we should read the value, synchronize and read the value again. OTOH I don't think that `multiple threads calling destroy(ContextState)` is something that should normally happen.

Reply:
Actually, I think that we could get rid of the synchronized block completely and use the `AtomicBoolean.compareAndSet()` instead... let's give it a try! :-)

Reply:
Used `VALUE.compareAndSet()`...